frontend/javascripts/dashboard/folders/metadata_table.tsx

```diff
@@ -0,0 +1,414 @@
+  const metadataWithoutIndexAndError = metadata.map(
+    ({ index: _ignored, error: _ignored2, ...rest }) => rest,
+  );
+  let serverResponse: APIDataset | Folder;
+  const isADataset = isDataset(datasetOrFolderToUpdate);
+  const datasetOrFolderString = isADataset ? "dataset" : "folder";
+  try {
+    if (isADataset) {
+      // In case of a dataset, update the dataset's metadata.
+      serverResponse = await context.updateCachedDataset(datasetOrFolderToUpdate, {
+        metadata: metadataWithoutIndexAndError,
+      });
+    } else {
+      // Else update the folders metadata.
+      const folder: Folder = datasetOrFolderToUpdate;
+      serverResponse = await context.queries.updateFolderMutation.mutateAsync({
+        ...folder,
+        allowedTeams: folder.allowedTeams?.map((t) => t.id) || [],
+        metadata: metadataWithoutIndexAndError,
+      });
+    }
+    if (!_.isEqual(serverResponse.metadata, metadataWithoutIndexAndError)) {
+      Toast.error(
+        `Failed to save metadata changes for ${datasetOrFolderString} ${datasetOrFolderToUpdate.name}.`,
+      );
+    } else {
+      !dontUpdateState && setHasUnsavedChanges(false);
+    }
+  } catch (error) {
+    Toast.error(
+      `Failed to save metadata changes for ${datasetOrFolderString} ${datasetOrFolderToUpdate.name}. Please look in the console for more details.`,
+    );
+    console.error(error);
+  } finally {
+    !dontUpdateState && setIsSaving(false);
+  }
+};
+
+const saveCurrentMetadataDebounced = _.debounce(saveCurrentMetadata, 3000);
+
+const getKeyInputId = (record: APIMetadataWithIndexAndError) =>
+  `metadata-key-input-id-${record.index}`;
+
+type APIMetadataWithIndexAndError = APIMetadata & { index: number; error?: string | null };
+type IndexedMetadataEntries = APIMetadataWithIndexAndError[];
+type EffectCancelSignal = { dontUpdateState: boolean };
+
+const isDataset = (datasetOrFolder: APIDataset | Folder): datasetOrFolder is APIDataset =>
+  "folderId" in datasetOrFolder;
+
+export default function MetadataTable({
+  datasetOrFolder,
+}: { datasetOrFolder: APIDataset | Folder }) {
+  const context = useDatasetCollectionContext();
+  const previousDatasetOrFolder: APIDataset | Folder | undefined =
+    usePreviousValue(datasetOrFolder);
+  const [metadata, setMetadata] = useState<IndexedMetadataEntries>(
+    datasetOrFolder?.metadata?.map((entry, index) => ({ ...entry, index, error: null })) || [],
+  );
+  const [focusedRow, setFocusedRow] = useState<number | null>(null);
+  const [isSaving, setIsSaving] = useState<boolean>(false);
+  const [hasUnsavedChanges, setHasUnsavedChanges] = useState<boolean>(false);
+
+  const isFirstRenderingCycle = previousDatasetOrFolder == null;
+
+  // Always update local state when folder / dataset changes. Prior to that send pending updates.
+  // biome-ignore lint/correctness/useExhaustiveDependencies: Only execute this hook when underlying datasetOrFolder changes.
+  useEffect(() => {
```

Comment:
not sure whether I understand it correctly: this metadata table is not remounted when activating another dataset/folder item in the list, right? therefore, you check whether the item changed in this effect. if yes, you flush the pending updates. is this correct? I think, it would be cleaner to unmount and remount the component when the dataset/folder changed. that way, the component's state is perfectly isolated (right now; I'd be a bit afraid that the metadata slips into the other dataset/folder in edge cases). the "use previous" code here wouldn't be necessary, either, I think.
you could achieve this easily by passing a `key` like this `<MetadataTable key={"${item.name}-${item.type}"} />`.
or was there a reason for not doing this?